juniper_actix_web/src/lib.rs

```diff
@@ -0,0 +1,620 @@
+    variables: Option<InputValue<S>>,
+}
+
+impl<S> Into<JuniperGraphQLRequest<S>> for StrictGraphQLRequest<S>
+where
+    S: ScalarValue,
+{
+    fn into(self) -> JuniperGraphQLRequest<S> {
+        JuniperGraphQLRequest::<S>::new(self.query, self.operation_name, self.variables)
+    }
+}
+
+impl<S> FromRequest for GraphQLRequest<S>
+where
+    S: ScalarValue + 'static,
+{
+    type Error = actix_web::Error;
+    type Future = Box<dyn Future<Item = Self, Error = Error>>;
+    type Config = ();
+
+    fn from_request(req: &HttpRequest, payload: &mut dev::Payload) -> Self::Future {
+        match req.method() {
+            &Method::GET => Box::new(future::result(
+                web::Query::<StrictGraphQLRequest<S>>::from_query(req.query_string())
+                    .map_err(Error::from)
+                    .map(|gql_request| {
+                        GraphQLRequest(GraphQLBatchRequest::Single(gql_request.into_inner().into()))
+                    }),
+            )),
+            &Method::POST => {
+                let content_type_header = req
+                    .headers()
+                    .get(actix_web::http::header::CONTENT_TYPE)
+                    .and_then(|hv| hv.to_str().ok());
+                match content_type_header {
+                    Some("application/json") => Box::new(
+                        web::Json::<GraphQLBatchRequest<S>>::from_request(req, payload)
+                            .map_err(Error::from)
+                            .map(|gql_request| GraphQLRequest(gql_request.into_inner())),
+                    ),
+                    Some("application/graphql") => Box::new(
+                        String::from_request(req, payload)
+                            .map_err(|_| actix_http::error::ContentTypeError::ParseError.into())
+                            .map(|query| {
+                                GraphQLRequest(GraphQLBatchRequest::Single(
+                                    JuniperGraphQLRequest::new(query, None, None),
+                                ))
+                            }),
+                    ),
+                    _ => Box::new(future::err(
+                        actix_http::error::ContentTypeError::UnknownEncoding.into(),
+                    )),
+                }
+            }
+            _ => Box::new(future::result(Err(
+                actix_http::error::ErrorMethodNotAllowed(
+                    "GraphQL requests can only be sent with GET or POST",
+                ),
+            ))),
+        }
+    }
+}
+
+impl Responder for GraphQLResponse {
+    type Error = actix_web::Error;
+    type Future = Result<HttpResponse, Error>;
+    fn respond_to(self, _: &HttpRequest) -> Self::Future {
+        let GraphQLResponse(status, body) = self;
+
+        Ok(HttpResponse::Ok()
+            .status(status)
+            .content_type("application/json")
+            .body(body))
+    }
+}
+
+#[cfg(test)]
+mod fromrequest_tests {
```

Comment:
Since you wrote quite a lot of tests ( :+1: ), let's factor them out to a separate file.